### PR TITLE
Includes mmctl into the mono-repo

### PR DESCRIPTION
**Pull request by mgdelacroix**
#### Summary
Adds the `mmctl` source code to the mono-repo and creates the necessary `make` targets for the tool to build and test correctly. Fixes some tests, makes the e2e ones pass with the products up and running and updates the compilation process set the tool's version from `model.CurrentVersion` and take the `BUILD_HASH` from the `Makefile`, so both match with the `mattermost` binary ones.

It is not expected to review the full `mmctl`s source as part of this PR, as only minor things have changed to adapt it to the main repo. Instead, I've left some comments pointing to the most significant changes, and in general, the things to check are the build pipeline, the CI and the `Makefile` targets.

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/build/release.mk [hunk 1]
```diff
@@ -156,16 +156,13 @@ package-general:
 	mkdir -p $(DIST_PATH_GENERIC)/logs
 	mkdir -p $(DIST_PATH_GENERIC)/prepackaged_plugins
 
-	@# Copy binary
+	@# Copy binaries
 ifeq ($(BUILDER_GOOS_GOARCH),"$(CURRENT_PACKAGE_ARCH)")
-	cp $(GOBIN)/$(MM_BIN_NAME) $(DIST_PATH_GENERIC)/bin # from native bin dir, not cross-compiled
+	cp $(GOBIN)/$(MM_BIN_NAME) $(GOBIN)/$(MMCTL_BIN_NAME) $(DIST_PATH_GENERIC)/bin # from native bin dir, not cross-compiled
```
Comment (mgdelacroix) [inline server/build/release.mk:161]: `build-cmd-*` targets are already building `./cmd/...`, so `mmctl` will be included in the `$(GOBIN)` without needing to update anything. We just need to remove the `scripts/download_mmctl_release.sh` call and make sure that the binary is copied over with the `mattermost` one for the build to include it

File: server/cmd/mmctl/commands/version.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	"runtime"
+
+	"github.com/mattermost/mattermost-server/server/v8/cmd/mmctl/printer"
+	"github.com/mattermost/mattermost-server/server/v8/model"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	Version = model.CurrentVersion
```
Comment (mgdelacroix) [inline server/cmd/mmctl/commands/version.go:17]: `mmctl` now reads its version directly from the `model` package

File: server/Makefile [hunk 3]
```diff
@@ -51,13 +58,24 @@ ifeq ($(BUILD_NUMBER),dev)
 	GOTAGS += "testlicensekey"
 endif
 
+# mmctl
+MMCTL_BUILD_TAGS =
+MMCTL_TESTFLAGS = -timeout 30m -race -v
+MMCTL_PKG = github.com/mattermost/mattermost-server/server/v8/cmd/mmctl/commands
+LDFLAGS += -X "$(MMCTL_PKG).gitCommit=$(BUILD_HASH)"
```
Comment (mgdelacroix) [inline server/Makefile:65]: We use directly `LDFLAGS` so we can build `./...` and `./cmd/...` later

**Comment by agnivade** [comment]
Nice work on this @mgdelacroix ! Excited to get this merged.

File: .github/workflows/test-mmctl.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+name: Test mmctl
```
Comment (isacikgoz) [inline .github/workflows/test-mmctl.yml:1]: To keep naming consistent, we can use the following convention:

```suggestion
name: mmctl CI
```

File: server/Makefile [hunk 5]
```diff
@@ -285,9 +305,9 @@ prepackaged-plugins: ## Populate the prepackaged-plugins directory
 
 prepackaged-binaries: ## Populate the prepackaged-binaries to the bin directory
 ifeq ($(shell test -f bin/mmctl && printf "yes"),yes)
-	@echo "MMCTL already exists in bin/mmctl not downloading a new version."
+	@echo "MMCTL already exists in bin/mmctl, not compiling."
 else
-	@scripts/download_mmctl_release.sh
+	$(MAKE) mmctl-build
```
Comment (isacikgoz) [inline server/Makefile:310]: This was long waited! :)

**Review by isacikgoz — COMMENTED** [review]
This looks good for my initial pass, I'm leaning on CI results for sure, are you planning to add checking docs step to the CI?

**Comment by mgdelacroix** [comment]
Added the docs check and updated the CI file structure to match the changes done for the rest of the jobs /cc @isacikgoz

File: server/Makefile [hunk 6]
```diff
@@ -51,13 +58,25 @@ ifeq ($(BUILD_NUMBER),dev)
 	GOTAGS += "testlicensekey"
 endif
 
+# mmctl
+MM_SERVER_PATH ?= $(PWD)
```
Comment (isacikgoz) [inline server/Makefile:62]: Not a blocker but what do you think if we there is a need of keeping this?

Comment (mgdelacroix) [reply server/Makefile:62]: Let me do a quick test to try and remove it. We definitely need to set it on the test targets (on server for example is done in the `scripts/test.sh` file, for `mmctl` we do it directly) but maybe we can simply use `$(PWD)` directly

**Review by isacikgoz — APPROVED** [review]
Nothing to block on my side, awesome work @mgdelacroix 🚀

**Comment by mgdelacroix** [comment]
Gentle ping @phoinixgrr

**Review by wiggin77 — APPROVED** [review]
LGTM :+1: Great work!

File: .github/workflows/mmctl-test-template.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    env:
+      COMPOSE_PROJECT_NAME: ghactions
+      BUILD_IMAGE: mattermost/mattermost-build-server:20230118_golang-1.19.5
+    steps:
+      - name: Checkout mattermost-server
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - name: Setup Go
+        uses: actions/setup-go@4d34df0c2316fe8122ab82dc22947d607c0c91f9 # v4.0.0
+        with:
+          go-version: ${{ env.go-version }}
```
Comment (phoinixgrr) [inline .github/workflows/mmctl-test-template.yml:25]: Is there a reason for hardcoding the go version? Setup-go can autodetect this using go.mod file. 

https://github.com/actions/setup-go#getting-go-version-from-the-gomod-file

**Review by phoinixgrr — APPROVED** [review]
Awesome work here 🎖️ 
Can't wait to have this merged

File: .github/workflows/mmctl-test-template.yml (hunk 7 above)
Comment (agnivade) [reply .github/workflows/mmctl-test-template.yml:25]: Yes, we need to pin it to the exact patch release version.

**Comment by hanzei** [comment]
Is this PR ready to be merged? If so, go for it. We can merge whatever PR is ready first.

**Comment by mgdelacroix** [comment]
@isacikgoz @hanzei I'm updating and retesting this PR, so it should be merged in the next hour

**Pull request merged**